
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
        <meta name="description" content="Lumi's Blog">
      
      
        <meta name="author" content="Lumi">
      
      
        <link rel="canonical" href="https://bhaaa-zlm.github.io/Bioinformatics/R/DESeq/">
      
      
        <link rel="prev" href="../../ATAC-Seq/ATAC-Seq_tools/">
      
      
        <link rel="next" href="../../../Biology/">
      
      
      <link rel="icon" href="../../../assets/image/fatLogo.png">
      <meta name="generator" content="mkdocs-1.6.0, mkdocs-material-9.5.27">
    
    
      
        <title>R and DESeq2 - Lumi's Blog</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.6543a935.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Libre+Baskerville:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Libre Baskerville";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../stylesheets/custom.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce((e,_)=>(e<<5)-e+_.charCodeAt(0),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="red" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#getting-ready-for-deseq2" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="Lumi&#39;s Blog" class="md-header__button md-logo" aria-label="Lumi's Blog" data-md-component="logo">
      
  <img src="../../../assets/image/fatLogo.png" alt="logo">

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2Z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            Lumi's Blog
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              R and DESeq2
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="default" data-md-color-primary="red" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 7a5 5 0 0 1 5 5 5 5 0 0 1-5 5 5 5 0 0 1-5-5 5 5 0 0 1 5-5m0 2a3 3 0 0 0-3 3 3 3 0 0 0 3 3 3 3 0 0 0 3-3 3 3 0 0 0-3-3m0-7 2.39 3.42C13.65 5.15 12.84 5 12 5c-.84 0-1.65.15-2.39.42L12 2M3.34 7l4.16-.35A7.2 7.2 0 0 0 5.94 8.5c-.44.74-.69 1.5-.83 2.29L3.34 7m.02 10 1.76-3.77a7.131 7.131 0 0 0 2.38 4.14L3.36 17M20.65 7l-1.77 3.79a7.023 7.023 0 0 0-2.38-4.15l4.15.36m-.01 10-4.14.36c.59-.51 1.12-1.14 1.54-1.86.42-.73.69-1.5.83-2.29L20.64 17M12 22l-2.41-3.44c.74.27 1.55.44 2.41.44.82 0 1.63-.17 2.37-.44L12 22Z"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="" data-md-color-scheme="slate" data-md-color-primary="red" data-md-color-accent="pink"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3 3.19.09m3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95 2.06.05m-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31Z"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var media,input,key,value,palette=__md_get("__palette");if(palette&&palette.color){"(prefers-color-scheme)"===palette.color.media&&(media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']"),palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent"));for([key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5Z"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12Z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41Z"/></svg>
        </button>
      </nav>
      
        <div class="md-search__suggest" data-md-component="search-suggest"></div>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/BHAAA-ZLM/BHAAA-ZLM.github.io" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    BHAAA-ZLM.github.io
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../Life/" class="md-tabs__link">
          
  
    
  
  Life

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../" class="md-tabs__link">
          
  
    
  
  Bioinformatics

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../Biology/" class="md-tabs__link">
          
  
    
  
  Biology

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../CS/" class="md-tabs__link">
          
  
    
  
  CS

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../Learning/" class="md-tabs__link">
          
  
    
  
  Learning

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../About/" class="md-tabs__link">
          
  
    
  
  About

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="Lumi&#39;s Blog" class="md-nav__button md-logo" aria-label="Lumi's Blog" data-md-component="logo">
      
  <img src="../../../assets/image/fatLogo.png" alt="logo">

    </a>
    Lumi's Blog
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/BHAAA-ZLM/BHAAA-ZLM.github.io" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg>
  </div>
  <div class="md-source__repository">
    BHAAA-ZLM.github.io
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../Life/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Life
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_1" id="__nav_1_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Life
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1_2" >
        
          
          <label class="md-nav__link" for="__nav_1_2" id="__nav_1_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Books I Read
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_1_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1_2">
            <span class="md-nav__icon md-icon"></span>
            Books I Read
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Life/Books/TAOS/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    The Art of Statistics
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
          
        
      
        
      
        
      
        
      
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" checked>
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Bioinformatics
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_2" id="__nav_2_label" tabindex="">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Bioinformatics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2_2" >
        
          
          <label class="md-nav__link" for="__nav_2_2" id="__nav_2_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    RNA-Seq
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_2">
            <span class="md-nav__icon md-icon"></span>
            RNA-Seq
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../RNA-seq/RNA-seq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    The Basic Workflow of RNA-seq
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../RNA-seq/Files/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Files involved in RNA-seq
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../RNA-seq/Tools/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    The Tools of RNA-seq
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_2_3" >
        
          
          <label class="md-nav__link" for="__nav_2_3" id="__nav_2_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    ATAC-Seq
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2_3">
            <span class="md-nav__icon md-icon"></span>
            ATAC-Seq
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../ATAC-Seq/ATAC-Seq_tools/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Tools in ATAC-Seq
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2_4" checked>
        
          
          <label class="md-nav__link" for="__nav_2_4" id="__nav_2_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    R the Programming Language
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_2_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_2_4">
            <span class="md-nav__icon md-icon"></span>
            R the Programming Language
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    R and DESeq2
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    R and DESeq2
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#getting-ready-for-deseq2" class="md-nav__link">
    <span class="md-ellipsis">
      Getting Ready for DESeq2
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Getting Ready for DESeq2">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#getting-the-matrix" class="md-nav__link">
    <span class="md-ellipsis">
      Getting the Matrix
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#input-the-matrix-into-r" class="md-nav__link">
    <span class="md-ellipsis">
      Input the Matrix into R
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#construct-the-deseq-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      Construct the DESeq Dataset
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#finally-the-deseq-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Finally, the DESeq analysis
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plot-your-data" class="md-nav__link">
    <span class="md-ellipsis">
      Plot Your Data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Plot Your Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-ma-plot" class="md-nav__link">
    <span class="md-ellipsis">
      The MA plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#heatmap" class="md-nav__link">
    <span class="md-ellipsis">
      Heatmap
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene-ontology" class="md-nav__link">
    <span class="md-ellipsis">
      Gene Ontology
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#volcano-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Volcano Plot
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../Biology/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Biology
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_3" id="__nav_3_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Biology
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_2" >
        
          
          <label class="md-nav__link" for="__nav_3_2" id="__nav_3_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Experiments
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_2">
            <span class="md-nav__icon md-icon"></span>
            Experiments
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Biology/Experiments/BIOS201/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Genome-Why are we different
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_3_3" >
        
          
          <label class="md-nav__link" for="__nav_3_3" id="__nav_3_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Structural Biology
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_3_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_3_3">
            <span class="md-nav__icon md-icon"></span>
            Structural Biology
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Biology/Struct/Chimera_visualization/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visualization with ChimeraX
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Biology/Struct/coot_install/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installing Coot on Mac
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../CS/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    CS
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_4" id="__nav_4_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            CS
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Learning CS
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            Learning CS
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../CS/Learn%20CS/Logic_review/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Learning Mathematical Logic
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../CS/Learn%20CS/Discrete_review/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Learning Discrete Mathematics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../CS/Learn%20CS/Markdown/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Learning Markdown
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../CS/Learn%20CS/R/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Learning R
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_3" >
        
          
          <label class="md-nav__link" for="__nav_4_3" id="__nav_4_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Projects
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_3">
            <span class="md-nav__icon md-icon"></span>
            Projects
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../CS/Projects/ChessKing/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    My First Project - ChessKing
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../Learning/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Learning
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_5" id="__nav_5_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            Learning
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    
    
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5_2" >
        
          
          <label class="md-nav__link" for="__nav_5_2" id="__nav_5_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Chemistry
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_2">
            <span class="md-nav__icon md-icon"></span>
            Chemistry
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Learning/Chemistry/Experiments/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Chemistry Experiments
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5_3" >
        
          
          <label class="md-nav__link" for="__nav_5_3" id="__nav_5_3_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Physics
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_5_3_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5_3">
            <span class="md-nav__icon md-icon"></span>
            Physics
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Learning/Physics/Quantum/Basics/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quantum Mechanics
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../Learning/Physics/fly_by_night/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Fly By Night
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
          
        
      
        
      
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_6" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../../About/" class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    About
  </span>
  

            </a>
            
              
              <label class="md-nav__link " for="__nav_6" id="__nav_6_label" tabindex="0">
                <span class="md-nav__icon md-icon"></span>
              </label>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_6_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_6">
            <span class="md-nav__icon md-icon"></span>
            About
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../About/SelfIntro/SelfIntro/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    About Me
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#getting-ready-for-deseq2" class="md-nav__link">
    <span class="md-ellipsis">
      Getting Ready for DESeq2
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Getting Ready for DESeq2">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#getting-the-matrix" class="md-nav__link">
    <span class="md-ellipsis">
      Getting the Matrix
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#input-the-matrix-into-r" class="md-nav__link">
    <span class="md-ellipsis">
      Input the Matrix into R
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#construct-the-deseq-dataset" class="md-nav__link">
    <span class="md-ellipsis">
      Construct the DESeq Dataset
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#finally-the-deseq-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Finally, the DESeq analysis
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#plot-your-data" class="md-nav__link">
    <span class="md-ellipsis">
      Plot Your Data
    </span>
  </a>
  
    <nav class="md-nav" aria-label="Plot Your Data">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#the-ma-plot" class="md-nav__link">
    <span class="md-ellipsis">
      The MA plot
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#heatmap" class="md-nav__link">
    <span class="md-ellipsis">
      Heatmap
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#gene-ontology" class="md-nav__link">
    <span class="md-ellipsis">
      Gene Ontology
    </span>
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#volcano-plot" class="md-nav__link">
    <span class="md-ellipsis">
      Volcano Plot
    </span>
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                

        <!-- Edit button -->
        
          <a
            href="https://github.com/BHAAA-ZLM/BHAAA-ZLM.github.io/edit/master/docs/Bioinformatics/R/DESeq.md"
            title="edit.link.title"
            class="md-content__button md-icon"
          >
            <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25Z"/></svg>
          </a>
        

        <!--
          Hack: check whether the content contains a h1 headline. If it
          doesn't, the page title (or respectively site name) is used
          as the main headline.
        -->
        
          <h1>R and DESeq2</h1>
        

        <!-- Markdown content -->
        
        <aside class="mdx-author">
            <p>
            <img src="https://avatars.githubusercontent.com/BHAAA-ZLM" alt="@BHAAA-ZLM">
            </p>
            <p>
            <span> 
            <b>Lumi</b> · <a href="https://github.com/BHAAA-ZLM">@BHAAA-ZLM</a>
            </span>
            <span>
                <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-calendar2" viewBox="0 0 16 16">
                    <path d="M3.5 0a.5.5 0 0 1 .5.5V1h8V.5a.5.5 0 0 1 1 0V1h1a2 2 0 0 1 2 2v11a2 2 0 0 1-2 2H2a2 2 0 0 1-2-2V3a2 2 0 0 1 2-2h1V.5a.5.5 0 0 1 .5-.5zM2 2a1 1 0 0 0-1 1v11a1 1 0 0 0 1 1h12a1 1 0 0 0 1-1V3a1 1 0 0 0-1-1H2z"/>
                    <path d="M2.5 4a.5.5 0 0 1 .5-.5h10a.5.5 0 0 1 .5.5v1a.5.5 0 0 1-.5.5H3a.5.5 0 0 1-.5-.5V4z"/>
                  </svg> 2022.8.3 ·
                  <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="currentColor" class="bi bi-clock" viewBox="0 0 16 16">
                    <path d="M8 3.5a.5.5 0 0 0-1 0V9a.5.5 0 0 0 .252.434l3.5 2a.5.5 0 0 0 .496-.868L8 8.71V3.5z"/>
                    <path d="M8 16A8 8 0 1 0 8 0a8 8 0 0 0 0 16zm7-8A7 7 0 1 1 1 8a7 7 0 0 1 14 0z"/>
                  </svg> 1hour read
            </span>
            </p>
        </aside>
        

        <p><span style="font-family: Courier"> DESeq2 is a very special R package made for performing differential expression analysis on your sequence, especially when you are tring to define differences between multiple biological conditions (e.g. treated, untreated).</p>
<p><span style="font-family: Courier">  Most of my overall knowledge comes from surfing the internet, and I found:</p>
<ul>
<li>
<p><span style="font-family: Courier"> A <a href="https://www.youtube.com/watch?v=OzNzO8qwwp0&amp;t=226s">wonderful girl</a> teaching DESeq2. Her channel is so rich in knowledge, everyone, subscribe now!!!</p>
</li>
<li>
<p><span style="font-family: Courier"> Another <a href="https://www.youtube.com/channel/UCuf90yPD_Yx53xZyVLtvRmA/videos">youtuber</a> who teaches all the different plots in both R and python.</p>
</li>
<li>
<p><span style="font-family: Courier"> The original <a href="http://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#countmat">manual</a> for DESeq2.</p>
</li>
<li>
<p><span style="font-family: Courier"> Another DESeq2 <a href="https://lashlock.github.io/compbio/R_presentation.html">tutorial</a> made very thoroughly and clear, I love it!</p>
</li>
<li>
<p><span style="font-family: Courier"> The DESeq2 <a href="https://genviz.org/module-04-expression/0004/02/01/DifferentialExpression/">tutorial</a> made by Griffith lab of Washington University. Which is surprisingly detailed when it comes to plotting.</p>
</li>
</ul>
<h2 id="getting-ready-for-deseq2"><span style="font-family: Courier"> Getting Ready for DESeq2</h2>
<p><span style="font-family: Courier">  In our RNA-seq workflow, we used cuffdiff to get a general differential expression result, from it, you can easily see which genes are expressed more, which are suppressed, in other words, which genes have a significant count difference from the other genes.</p>
<p><span style="font-family: Courier"> But when we come to the real biological experiments and analysis, we usually want to know more about our gene and about our test subjects. We might even want to draw fancy pictures to convey our ideas. That's when our rather old tool cuffdiff doesn't shine. </p>
<p><span style="font-family: Courier"> DESeq2 basically work nearly the same as cuffdiff (although the algorithms might be a bit different), but the final result we are looking for are bascially the same. The log2(foldchanges) and p_values and all those things. So how does it work?</p>
<h3 id="getting-the-matrix"><span style="font-family: Courier"> Getting the Matrix</h3>
<p><span style="font-family: Courier"> First we need the count matrix from our data. We need an old friend from our RNA-seq workflow - featureCounts. </p>
<pre><code>featureCounts [options] -a &lt;annotation_file&gt; -o &lt;output_file&gt; input_file1 [input_file2] ... 
</code></pre>
<p><span style="font-family: Courier"> By typing this command into the terminal, feature counts will automatically map all the counts to the genes in your annotation file. Thus providing you with a matrix of counts for specific genes. We only need the counts for each matrix, so we can cut the matrix in the command line or in R. We might also want to remove some data with very low counts, since they are not very significant, and we can do that also in the command line or in R.</p>
<h3 id="input-the-matrix-into-r"><span style="font-family: Courier"> Input the Matrix into R</h3>
<p><span style="font-family: Courier"> The next step is to get your matrix into R and prepare it for DESeq2. The code I used was this, and I found it pretty elegant.</p>
<pre><code class="language-R"># input data 
input_data &lt;- read.table(&quot;deseq2_input.txt&quot;,header = T, row.names = 1)


# preperation 
input_data &lt;- as.matrix(input_data)
condition &lt;- factor(c(rep(&quot;ctrl&quot;,4), rep(&quot;DDX6KO&quot;,4)))
coldata &lt;- data.frame(row.names = colnames(input_data),condition)
</code></pre>
<p><span style="font-family: Courier"> First we input the data into the variable input_data by the R function read.table(). You can always get help in R by just typing <code>?functionName()</code> in the R console. So in the read.table() function, we set the first row as our header with <code>header = T</code>. <code>row.names = 1</code> indicates that the first column of each row gives the name of the first row.</p>
<p><span style="font-family: Courier"> Then, we prepare the raw data so that it fits the requirements of DESeq2. Apart from the original data matrix from your sequencing results, DESeq2 also requires another matrix containing informations on the columns of our data (in other words, the names and types of your contol and experiment groups). So the <code>as.matrix</code> turns the data into a matrix fit for calculation. <code>condition</code> stores a factor of two (it can be more than two). <code>coldata &lt;- data.frame()</code> gives DESeq2 the column data it needs.</p>
<h3 id="construct-the-deseq-dataset"><span style="font-family: Courier"> Construct the DESeq Dataset</h3>
<p><span style="font-family: Courier"> Then we use our data to construct a DESeq Dataset from our matrix (in this case, <code>input_data</code>).</p>
<pre><code class="language-R">library(DESeq2)
# construct a deseq input data matrix
dds &lt;- DESeqDataSetFromMatrix(countData = input_data, colData = coldata, design = ~condition)

dim(dds)
dim(dds[rowSums(counts(dds)) &gt; 5,])
dds &lt;- dds[rowSums(counts(dds)) &gt; 5,]
</code></pre>
<p><span style="font-family: Courier"> We first need to import the DESeq2 library (package) (of course). </p>
<p><span style="font-family: Courier"> Then, we use the function <code>DESeqDataSetFromMatrix()</code> from DESeq2 package to input our raw data. <code>colData</code> is the matrix we made earlier containing the data for row names and types and all those stuff. <code>design = ~condtion</code> means that the data is designed by the factor <code>condition</code> in the <code>colData</code> variable. In designed, I mean that the DESeq process will perform the differential expression analysis based on the different conditions of the experiments. The conditions are also set manually. You can add another column in the <code>colData</code> matrix called <code>time</code>, and set the design to <code>design = ~condition + time</code>, DESeq will restrain the effects of condition and focus on the time factor. For more information, see <a href="https://www.biostars.org/p/278684/">this blog</a> on BioStars.</p>
<p><span style="font-family: Courier">  We can check the dimensions of dds using <code>dim(dds)</code>. We could remove the results when the gene's counts are less than 5 in total, by reassigning the value of dds to <code>dds[rowSums(counts(dds)) &gt; 5,]</code>. Thus we got our DESeq Dataset, named <code>dds</code>, and we can happily perform differential expression analysis on it!~</p>
<h3 id="finally-the-deseq-analysis"><span style="font-family: Courier"> Finally, the DESeq analysis</h3>
<p><span style="font-family: Courier"> The actual analysis part is really easy to understand, and doesn't require a lot of code of brain.</p>
<pre><code class="language-R"># perform the differential expression analysis
dds &lt;- DESeq(dds)
res &lt;- results(dds)
</code></pre>
<p><span style="font-family: Courier"> Just type these words and everything will be done in the background thanks to our lovely DESeq2 developers. The deeper truth behind all the analysis is much more complex, and I don't think I fully understand it. I will leave a link here to the <a href="https://www.youtube.com/watch?v=0b24mpzM_5M">lovely Indian girl's video</a>, and I might make another whole chapter just talking about how it works. But for us who are just using it and only cares about the final output data, these two lines of code is enough.</p>
<h2 id="plot-your-data"><span style="font-family: Courier"> Plot Your Data</h2>
<p><span style="font-family: Courier"> After we obtained our data from DESeq, we need to know how to show everyone else in a clear and fasionable way, what is happening in our experiments. That's where our beautiful plots come. A wide variaty of plots can be used to express our data, and it's best if we know how to draw each one of them, and more importantly, the essence behind every plot.</p>
<p><span style="font-family: Courier"> The knowledge of the plotting part mainly comes from the Griffith lab website, which is <a href="https://genviz.org/module-04-expression/0004/02/01/DifferentialExpression/">here</a>. There might be some websites helpful for individual plotting, I will list each of the websites in their own part.</p>
<h3 id="the-ma-plot"><span style="font-family: Courier"> The MA plot</h3>
<p><span style="font-family: Courier"> <a href="https://en.wikipedia.org/wiki/MA_plot">MAplots</a> are plots that can display the difference of measurements between two sets of data.</p>
<p><span style="font-family: Courier"> DESeq2 have it's own MAplotting function. Which is simply:</p>
<pre><code class="language-R">plotMA(res, ylim = c(-3,3))
</code></pre>
<p><img src="https://github.com/BHAAA-ZLM/BHAAA-ZLM.github.io/blob/main/docs/Bioinformatics/R/DEseq/plotMA.png?raw=true" width=600 tab="MAplot from DESeq2" title="MAplot from DESeq2"></p>
<p><span style="font-family: Courier"> But this MA plot only have two colours, and is rather lame. With ggplot's function geom_point, we could make a similar MA plot.</p>
<pre><code class="language-R">library(ggplot2)
deseq2ResDF &lt;- as.data.frame(res) ## convert the result file into a data frame
deseq2ResDF$significant &lt;- ifelse(deseq2ResDF$padj &lt; .1, &quot;Significant&quot;, NA) ## a new column for significance

ggplot(deseq2ResDF, aes(baseMean, log2FoldChange, colour = padj)) + 
  geom_point(size=0.5) + scale_y_continuous(limits=c(-3, 3)) + 
  scale_x_log10() + geom_hline(yintercept = 0, colour=&quot;darkorchid4&quot;, size=1, linetype=&quot;longdash&quot;) + 
  labs(x=&quot;mean of normalized counts&quot;, y=&quot;log fold change&quot;) + 
  theme_bw() + geom_density_2d(colour=&quot;black&quot;, size=0.25) +
  scale_colour_gradient( high = &quot;#f8ff2b&quot;, low = &quot;#ff2b2b&quot;, space = &quot;Lab&quot;, na.value = &quot;grey50&quot;, guide = &quot;colourbar&quot;, aesthetics = &quot;colour&quot;)
</code></pre>
<p><img src="https://github.com/BHAAA-ZLM/BHAAA-ZLM.github.io/blob/main/docs/Bioinformatics/R/DEseq/ggplotMA2.png?raw=true" width=600 tab="MAplot from ggplot" title="MAplot from ggplot"></p>
<p><span style="font-family: Courier">  By adding a series of constraints, we can make the geom_point look just like our MAplot. But first, we need to input our DESeq Data as a dataframe, and than add a new column called significant to the end of the dataframe. Remember to add <code>scale_x_log10</code> or else it will display the original counts, which are massive.</p>
<h3 id="heatmap"><span style="font-family: Courier"> Heatmap</h3>
<p><span style="font-family: Courier"> The heatmap knowledge comes from this <a href="https://www.youtube.com/watch?v=ht1r34-ifVI">video</a>.</p>
<p><span style="font-family: Courier"> Heatmaps are often used to show the difference between two groups of data, for example the experiment and the control. We can use the data we got from DESeq to plot a wonderful heatmap.</p>
<p><img src="https://github.com/BHAAA-ZLM/BHAAA-ZLM.github.io/blob/main/docs/Bioinformatics/R/DEseq/ComplexHeatMap.png?raw=true" width=600 tab="a complex heatmap" title="a complex heatmap"></p>
<p><span style="font-family: Courier"> A fantastic heatmap tutorial can be found <a href="https://www.youtube.com/watch?v=ht1r34-ifVI&amp;list=PLi1VnGoeDGjvHvl83QySD2oAQYFHPRYso&amp;index=8">here</a>. </p>
<pre><code class="language-R">library(DESeq2)
library(ComplexHeatmap)
library(RColorBrewer)
library(circlize)
</code></pre>
<p><span style="font-family: Courier"> Four libraries are used to draw this kind of heatmap. </p>
<pre><code class="language-R">dds_sig &lt;- subset(deseq2ResDF, significant == &quot;Significant&quot;)
dds_sig &lt;- dds_sig[(dds_sig$baseMean &gt; 50) &amp; (abs(dds_sig$log2FoldChange) &gt; 2),]
dds_sig &lt;- dds_sig[order(dds_sig$log2FoldChange, decreasing = T),]  
</code></pre>
<p><span style="font-family: Courier"> We first make a subset of our result matrix based on the significance (which comes from the padj number, usually padj &lt;.1 can be considered significant). Then we filter this data by deleting the rows that have very small baseMean number (which means it's probably background noise), and choose those genes that have the largest foldchange (which are the most significant genes). Then we order our data so that the genes that have the largest change are at the first and last rows of the matrix.</p>
<pre><code class="language-R">rlog_out &lt;- rlog(dds, blind = F)
mat &lt;- assay(rlog_out)[rownames(dds_sig),rownames(coldata)]
colnames(mat) &lt;- rownames(coldata)
base_mean &lt;- rowMeans(mat)
</code></pre>
<p><span style="font-family: Courier"> Then we perform the rlog function on the DESeq Dataset to calculate the <a href="https://en.wikipedia.org/wiki/Standard_score">Z-score</a> for the genes. Because the rlog_out is another dataset, we transform it into another matrix and set the column and row name of the matrix.</p>
<pre><code class="language-R">mat_scaled &lt;- t(apply(mat, 1, scale))
colnames(mat_scaled) &lt;- colnames(mat)
</code></pre>
<p><span style="font-family: Courier">  Then we <a href="https://en.wikipedia.org/wiki/Scaling_(geometry)">scale</a> the matrix, which I guess is similar to the SVD decomposition, to get their influence factors on the difference. In our case, the scaling is more like normalizing the data of each row of mat, thus setting a global standard to express the difference between each group. Then set the names of our new matrix, and then we are done. </p>
<pre><code class="language-R">rows_keep &lt;- c(seq(1:25), seq((nrow(mat_scaled) - 25), nrow(mat_scaled)))

l2_val &lt;- as.matrix(dds_sig[rows_keep,]$log2FoldChange)
colnames(l2_val) &lt;- &quot;logFC&quot;

mean_val &lt;- as.matrix(dds_sig[rows_keep,]$baseMean)
colnames(mean_val) &lt;- &quot;AveExpr&quot;
</code></pre>
<p><span style="font-family: Courier"> Then we chose the 50 genes that have the largest and smallest foldchange to plot. We also want to plot their log2FoldChange value and baseMean value, so we extract them from the original sorted data (not from the normalized matrix).</p>
<pre><code class="language-R">col_logFC &lt;- colorRamp2(c(min(l2_val),0,max(l2_val)),c(&quot;blue&quot;,&quot;white&quot;,&quot;red&quot;))
col_AveExpr &lt;- colorRamp2(c(quantile(mean_val)[1], quantile(mean_val)[4]), c(&quot;white&quot;,&quot;red&quot;))
</code></pre>
<p><span style="font-family: Courier"> We set the colour for our log(FoldChange) and baseMean sub-heatmap so that it matches our main matrix.</p>
<pre><code class="language-R">h1 &lt;- Heatmap(mat_scaled[rows_keep,], cluster_columns = T, cluster_rows = F,
              column_labels = colnames(mat_scaled), name = &quot;Z-score&quot;)

ha &lt;- HeatmapAnnotation(summary = anno_summary(gp = gpar(fill = 2), height = unit(2, &quot;cm&quot;)))
h2 &lt;- Heatmap(l2_val, row_labels = rownames(dds_sig[rows_keep,]),
              cluster_rows = F, name = &quot;LogFC&quot;, top_annotation = ha,
              col = col_logFC, 
              cell_fun = function(j, i, x, y, w, h, col){
                grid.text(round(l2_val[ i, j], 2), x, y)
              })

h3 &lt;- ComplexHeatmap::Heatmap(mean_val, row_labels = rownames(dds_sig[rows_keep,]),
              cluster_rows = F, name = &quot;AveExpr&quot;, 
              col = col_AveExpr, 
              cell_fun = function(j,i,x,y,w,h,col){
                grid.text(round(mean_val[ i, j], 2), x, y)
              })

h &lt;- h1 + h2 + h3
h
</code></pre>
<p><span style="font-family: Courier"> Then we finally plot the heatmap. <code>h1</code> is the main heatmap which demonstrates the expression difference of each gene in different group. <code>h2</code> and <code>ha</code> are bound together, <code>h2</code> shows the fold change of each individual gene, and <code>ha</code> is a small heatmap annotation showing the range of log(foldChange). At last <code>h3</code> is the baseMean value heatmap. We then combine all three graphs together, and get our final complex heatmap.</p>
<h3 id="gene-ontology"><span style="font-family: Courier"> Gene Ontology</h3>
<p><span style="font-family: Courier"> The gene ontology knowledge comes from this <a href="https://www.youtube.com/watch?v=JPwdqdo_tRg">video</a>.</p>
<pre><code class="language-R"># BiocManager::install(&quot;AnnotationDbi&quot;)
# BiocManager::install
library(clusterProfiler)
library(AnnotationDbi)
library(org.Hs.eg.db)
</code></pre>
<p><span style="font-family: Courier"> The three R packages we need to plot a gene ontology mapping.</p>
<pre><code class="language-R">gene_up &lt;- rownames(dds_sig[dds_sig$log2FoldChange &gt; 1.5,])
</code></pre>
<p><span style="font-family: Courier"> Filter out the genes that have the log2FoldChange bigger than 1.5. Which are the most effected pathways.</p>
<pre><code class="language-R">GO_up &lt;- enrichGO(gene = gene_up, OrgDb = &quot;org.Hs.eg.db&quot;, keyType = &quot;SYMBOL&quot;, ont = &quot;BP&quot;)
</code></pre>
<p><span style="font-family: Courier"> We then perform the gene ontology analysis on our genes, the gene ontology package can match each gene to their relating pathway. Our gene is a human gene, so we use the human reference, thus the <code>org.Hs.eg.db</code> (if we need the mouse reference, we can use <code>org.Mm.eg.db</code>). <code>keyType</code> might be ENSEMBL or SYMBOL, depending on the type of your gene's name. For <code>ont</code> we usually choose BP.</p>
<p><img src="https://github.com/BHAAA-ZLM/BHAAA-ZLM.github.io/blob/main/docs/Bioinformatics/R/DEseq/GOtypes.png?raw=true" width=600 title="3 GO type" alt="3 GO type"></p>
<pre><code class="language-R">up_plot &lt;- plot(barplot(GO_up, showCategory = 20))
up_plot
</code></pre>
<p><span style="font-family: Courier"> Then we plot the first 20 pathways, and thus show what pathways our experiment influence most. The graph looks something like this. We can make it look more beautiful by adjusting different variables.</p>
<p><img src="https://github.com/BHAAA-ZLM/BHAAA-ZLM.github.io/blob/main/docs/Bioinformatics/R/DEseq/GO.png?raw=true" width=600 title="GO plot" alt="GO plot"></p>
<h3 id="volcano-plot"><span style="font-family: Courier"> Volcano Plot</h3>
<p><span style="font-family: Courier"> The volcano plot knowledge comes from this <a href="https://www.youtube.com/watch?v=vRr78s37CI4">video</a>.</p>
<p><span style="font-family: Courier"> Volcano Plots can show you the differentially expressed genes, using the EnhancedVolcano library, we can easily plot amazing volcano plots easily.</p>
<pre><code class="language-R">res_vol &lt;- as.data.frame(res0.01[res0.01$baseMean &gt; 200, ])

library(EnhancedVolcano)

sel_tab &lt;- c(&quot;DDX6&quot;,&quot;IFI27&quot;,&quot;HPGD&quot;,&quot;GPI&quot;,&quot;PRF1&quot;)

EnhancedVolcano(res_vol, x = &quot;log2FoldChange&quot;, y = &quot;padj&quot;, lab = rownames(res_vol),
                selectLab = sel_tab, title = &quot;DDX6KO&quot;, subtitle = &quot;&quot;)
</code></pre>
<p><span style="font-family: Courier"> First we filter the low count data, and pass it into a dataframe. We could plot out the selected gene tabs using the variable <code>sel_tab</code>.</p>
<p><img src="https://github.com/BHAAA-ZLM/BHAAA-ZLM.github.io/blob/main/docs/Bioinformatics/R/DEseq/DDX6volPlot.png?raw=true" title="volcano plot" tab="volcano plot" width=600></p>

        <!-- Last update of source file -->
        
          
        
      
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12Z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright &copy; 2022-2024 Lumi
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
        <div class="md-social">
  
    
    
    
    
    <a href="https://space.bilibili.com/85744189" target="_blank" rel="noopener" title="Bilibili | MrBHAAA" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M488.6 104.1c16.7 18.1 24.4 39.7 23.3 65.7v202.4c-.4 26.4-9.2 48.1-26.5 65.1-17.2 17-39.1 25.9-65.5 26.7H92.02c-26.45-.8-48.21-9.8-65.28-27.2C9.682 419.4.767 396.5 0 368.2V169.8c.767-26 9.682-47.6 26.74-65.7C43.81 87.75 65.57 78.77 92.02 78h29.38L96.05 52.19c-5.75-5.73-8.63-13-8.63-21.79 0-8.8 2.88-16.06 8.63-21.797C101.8 2.868 109.1 0 117.9 0s16.1 2.868 21.9 8.603L213.1 78h88l74.5-69.397C381.7 2.868 389.2 0 398 0c8.8 0 16.1 2.868 21.9 8.603 5.7 5.737 8.6 12.997 8.6 21.797 0 8.79-2.9 16.06-8.6 21.79L394.6 78h29.3c26.4.77 48 9.75 64.7 26.1zm-38.8 69.7c-.4-9.6-3.7-17.4-10.7-23.5-5.2-6.1-14-9.4-22.7-9.8H96.05c-9.59.4-17.45 3.7-23.58 9.8-6.14 6.1-9.4 13.9-9.78 23.5v194.4c0 9.2 3.26 17 9.78 23.5s14.38 9.8 23.58 9.8H416.4c9.2 0 17-3.3 23.3-9.8 6.3-6.5 9.7-14.3 10.1-23.5V173.8zm-264.3 42.7c6.3 6.3 9.7 14.1 10.1 23.2V273c-.4 9.2-3.7 16.9-9.8 23.2-6.2 6.3-14 9.5-23.6 9.5-9.6 0-17.5-3.2-23.6-9.5-6.1-6.3-9.4-14-9.8-23.2v-33.3c.4-9.1 3.8-16.9 10.1-23.2 6.3-6.3 13.2-9.6 23.3-10 9.2.4 17 3.7 23.3 10zm191.5 0c6.3 6.3 9.7 14.1 10.1 23.2V273c-.4 9.2-3.7 16.9-9.8 23.2-6.1 6.3-14 9.5-23.6 9.5-9.6 0-17.4-3.2-23.6-9.5-7-6.3-9.4-14-9.7-23.2v-33.3c.3-9.1 3.7-16.9 10-23.2 6.3-6.3 14.1-9.6 23.3-10 9.2.4 17 3.7 23.3 10z"/></svg>
    </a>
  
    
    
    
    
    <a href="https://github.com/BHAAA-ZLM" target="_blank" rel="noopener" title="GitHub | Lumi" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 496 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg>
    </a>
  
    
    
    
    
    <a href="https://x.com/BHAAA2021374718" target="_blank" rel="noopener" title="X(twitter) | Lumi" class="md-social__link">
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.5.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M389.2 48h70.6L305.6 224.2 487 464H345L233.7 318.6 106.5 464H35.8l164.9-188.5L26.8 48h145.6l100.5 132.9L389.2 48zm-24.8 373.8h39.1L151.1 88h-42l255.3 333.8z"/></svg>
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.tabs", "navigation.expand", "navigation.top", "navigation.expand", "navigation.indexes", "search.suggest", "search.highlight"], "search": "../../../assets/javascripts/workers/search.b8dbb3d2.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.ad660dcc.min.js"></script>
      
        <script src="../../../javascripts/mathjax.js"></script>
      
        <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
      
        <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
      
    
  </body>
</html>